[DebugInfo] Update DIBuilder insertion to take InsertPosition

hvdijk opened the pull request:
After #124287 updated several functions to return iterators rather than Instruction *, it was no longer straightforward to pass their result to DIBuilder. This commit updates DIBuilder methods to accept an InsertPosition instead, so that they can be called with an iterator (preferred), or with a deprecation warning an Instruction *, or a BasicBlock *. This commit also updates the existing calls to the DIBuilder methods to pass in iterators.

jmorse commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -1708,19 +1706,9 @@ static void insertDbgValueOrDbgVariableRecord(DIBuilder &Builder, Value *DV,
-        new DbgVariableRecord(DVAM, DIVar, DIExpr, NewLoc.get());
-    Instr->getParent()->insertDbgRecordAfter(DV, &*Instr);
-  }
+    const DebugLoc &NewLoc, Instruction *Instr) {
+  insertDbgValueOrDbgVariableRecord(Builder, DV, DIVar, DIExpr, NewLoc,
+                                    std::next(Instr->getIterator()));
```
I endorse the amount of red here; I believe we need to call `setHeadBit(true)` on the next-iterator though, to ensure that in sequences such as:

    inst
    #dbg_value
    inst2

Inserting after `inst` will come before the pre-existing dbg_value, rather than after it.

hvdijk replied:
I think you're right, thanks, will update and re-test.

jmorse reviewed (commented):
This looks great, thanks for putting this all together, one nit inline,

Is this intended to be cherry-picked into the release branch or not? Seeing how the original motivation was to ease downstream compatibilities, I'd assumed there'd be the need for something to be cherry-picked, and this might be too invasive at this stage to be moved across. I don't have a full understanding of the use case though.

@OCHyams I believe you did the C API changes, are there any additional concerns in this area?

hvdijk commented:
> Is this intended to be cherry-picked into the release branch or not?

Yes, I was hoping that we were still early enough to be able to cherry-pick it without issues, especially as the aim is to restore a certain level of compatibility with previous LLVM releases. If that is not the case though, then at least it will be helpful to know that only LLVM 20 will require workarounds.

hvdijk committed:
[DebugInfo] Update DIBuilder insertion to take InsertPosition

After #124287 updated several functions to return iterators rather than
Instruction *, it was no longer straightforward to pass their result to
DIBuilder. This commit updates DIBuilder methods to accept an
InsertPosition instead, so that they can be called with an iterator
(preferred), or with a deprecation warning an Instruction *, or a
BasicBlock *. This commit also updates the existing calls to the
DIBuilder methods to pass in iterators.

OCHyams commented on llvm/lib/IR/DebugInfo.cpp:
```diff
@@ -1686,7 +1686,8 @@ LLVMDbgRecordRef LLVMDIBuilderInsertDeclareRecordBefore(
       unwrap(Storage), unwrap<DILocalVariable>(VarInfo),
       unwrap<DIExpression>(Expr), unwrap<DILocation>(DL),
-      unwrap<Instruction>(Instr));
+      Instr ? InsertPosition(unwrap<Instruction>(Instr)->getIterator())
+            : nullptr);
```
Wouldn't this trip an assertion in the `nullptr` InsertPosition case (`assert(InsertPt.isValid());` in `insertDbgVariableRecord` called by `insertDeclare`). Or am I reading this wrong? (Same question applies to the other calls below)

OCHyams replied:
Thanks for unpicking that.

> This is a messy situation and in this PR I tried to just keep that exactly the same. What previously worked should still work. What previously didn't work should still not work.

SGTM. I feel as though these functions should probably have asserts in them saying as such but that could reasonably be argued to be out of scope of the patch.

Patch LGTM to me too.

hvdijk replied:
Thanks! I think for consistency, it would be nice to update these functions so that they all treat `nullptr` consistently, as a request to create whatever the function says to create but not insert it anywhere. When I have some extra time I'll see if that's easy to do.

OCHyams reviewed (commented):
> @OCHyams I believe you did the C API changes, are there any additional concerns in this area?

SGTM, just one inline question

jmorse reviewed (approved):
LGTM too; thanks for running with this!

hvdijk commented:
The buildbot failures look like something conflicted with changes that have since gone in on LLVM. I have reverted the change for now and will check what is going on and update.
